diff --git a/openmp/runtime/cmake/LibompDefinitions.cmake b/openmp/runtime/cmake/LibompDefinitions.cmake
index 1d8a88966d939..c85e15323e0ee 100644
--- a/openmp/runtime/cmake/LibompDefinitions.cmake
+++ b/openmp/runtime/cmake/LibompDefinitions.cmake
@@ -16,7 +16,12 @@ function(libomp_get_definitions_flags cppflags)
     libomp_append(cppflags_local "-D _CRT_SECURE_NO_DEPRECATE")
     libomp_append(cppflags_local "-D _WINDOWS")
     libomp_append(cppflags_local "-D _WINNT")
-    libomp_append(cppflags_local "-D _WIN32_WINNT=0x0501")
+    if (MSVC)
+      # Force a default target OS version with MSVC based toolchains.
+      # (For MinGW based ones, use the toolchain's default target or what
+      # the user set in CMake flags.)
+      libomp_append(cppflags_local "-D _WIN32_WINNT=0x0501")
+    endif()
     libomp_append(cppflags_local "-D _USRDLL")
     libomp_append(cppflags_local "-D _ITERATOR_DEBUG_LEVEL=0" IF_TRUE DEBUG_BUILD)
     libomp_append(cppflags_local "-D _DEBUG" IF_TRUE DEBUG_BUILD)
