diff --git a/flang/test/Lower/OpenMP/Todo/allocate-clause-allocator.f90 b/flang/test/Lower/OpenMP/Todo/allocate-clause-allocator.f90
index d33bdf9f9a8db..90158a61f7503 100644
--- a/flang/test/Lower/OpenMP/Todo/allocate-clause-allocator.f90
+++ b/flang/test/Lower/OpenMP/Todo/allocate-clause-allocator.f90
@@ -4,7 +4,8 @@
 ! CHECK: not yet implemented: Unhandled clause allocate in omp.parallel
 ! CHECK: LLVM Translation failed for operation: omp.parallel
 program p
-  use omp_lib
+  !use omp_lib
+  integer(8),parameter::omp_default_mem_alloc=1_8
   integer :: x
   integer :: a
   integer :: i
