diff --git a/flang/lib/Lower/OpenMP/Clauses.cpp b/flang/lib/Lower/OpenMP/Clauses.cpp
index 9fa9abd9e8ceb..86d78e0358677 100644
--- a/flang/lib/Lower/OpenMP/Clauses.cpp
+++ b/flang/lib/Lower/OpenMP/Clauses.cpp
@@ -165,6 +165,7 @@ std::optional<Object> getBaseObject(const Object &object,
       // e.g. A(i) is represented as {Symbol(A), Designator(ArrayRef(A, i))}.
       // Here we have the Symbol(A), which is what we started with.
       assert(&**symRef == object.sym());
+      [[maybe_unused]] auto *unused = symRef;
       return std::nullopt;
     }
   } else {
