diff --git a/openmp/runtime/src/z_Linux_util.cpp b/openmp/runtime/src/z_Linux_util.cpp
index 0c89f2346750f..87f2711186420 100644
--- a/openmp/runtime/src/z_Linux_util.cpp
+++ b/openmp/runtime/src/z_Linux_util.cpp
@@ -2520,6 +2520,8 @@ int __kmp_get_load_balance(int max) {
   glb_running_threads = running_threads;
 
   return running_threads;
+}
+
 #elif KMP_OS_HAIKU
 
 int __kmp_get_load_balance(int max) { return -1; }
