diff --git a/openmp/runtime/src/ompt-internal.h b/openmp/runtime/src/ompt-internal.h
index 0cfab8bfaa190..36b45f7a91ea2 100644
--- a/openmp/runtime/src/ompt-internal.h
+++ b/openmp/runtime/src/ompt-internal.h
@@ -109,7 +109,8 @@ void ompt_pre_init(void);
 void ompt_post_init(void);
 void ompt_fini(void);
 
-#define OMPT_GET_RETURN_ADDRESS(level) __builtin_return_address(level)
+#define OMPT_GET_RETURN_ADDRESS(level)                                         \
+  __builtin_extract_return_addr(__builtin_return_address(level))
 #define OMPT_GET_FRAME_ADDRESS(level) __builtin_frame_address(level)
 #define OMPT_FRAME_FLAGS_APP (ompt_frame_application | ompt_frame_cfa)
 #define OMPT_FRAME_FLAGS_RUNTIME (ompt_frame_runtime | ompt_frame_cfa)
diff --git a/openmp/runtime/src/ompt-specific.h b/openmp/runtime/src/ompt-specific.h
index e9e40d43429ea..b7eb140458b40 100644
--- a/openmp/runtime/src/ompt-specific.h
+++ b/openmp/runtime/src/ompt-specific.h
@@ -102,15 +102,16 @@ inline void *__ompt_load_return_address(int gtid) {
   if (ompt_enabled.enabled && gtid >= 0 && __kmp_threads[gtid] &&              \
       !__kmp_threads[gtid]->th.ompt_thread_info.return_address)                \
   __kmp_threads[gtid]->th.ompt_thread_info.return_address =                    \
-      __builtin_return_address(0)*/
+      __builtin_extract_return_addr(__builtin_return_address(0))*/
 #define OMPT_STORE_RETURN_ADDRESS(gtid)                                        \
-  OmptReturnAddressGuard ReturnAddressGuard{gtid, __builtin_return_address(0)};
+  OmptReturnAddressGuard ReturnAddressGuard{                                   \
+      gtid, __builtin_extract_return_addr(__builtin_return_address(0))};
 #define OMPT_LOAD_RETURN_ADDRESS(gtid) __ompt_load_return_address(gtid)
 #define OMPT_LOAD_OR_GET_RETURN_ADDRESS(gtid)                                  \
   ((ompt_enabled.enabled && gtid >= 0 && __kmp_threads[gtid] &&                \
     __kmp_threads[gtid]->th.ompt_thread_info.return_address)                   \
        ? __ompt_load_return_address(gtid)                                      \
-       : __builtin_return_address(0))
+       : __builtin_extract_return_addr(__builtin_return_address(0)))
 
 #define OMPT_GET_DISPATCH_CHUNK(chunk, lb, ub, incr)                           \
   do {                                                                         \
