diff --git a/openmp/libompd/test/lit.site.cfg.in b/openmp/libompd/test/lit.site.cfg.in
index 63b8c97637d61..52e63f48bdc65 100644
--- a/openmp/libompd/test/lit.site.cfg.in
+++ b/openmp/libompd/test/lit.site.cfg.in
@@ -12,10 +12,14 @@ config.library_dir = "@LIBOMP_LIBRARY_DIR@"
 config.ompd_library_dir = "@CMAKE_CURRENT_BINARY_DIR@/../src/"
 config.omp_header_directory = "@LIBOMP_BINARY_DIR@/src"
 config.operating_system = "@CMAKE_SYSTEM_NAME@"
+config.target_triple = "@LLVM_TARGET_TRIPLE@"
 
 config.ompt_plugin = "@OMPT_PLUGIN@"
 config.ompt_include_dir = "@LIBOMP_INCLUDE_DIR@"
 config.ompd_module = "@CMAKE_CURRENT_BINARY_DIR@/../gdb-plugin/python-module/ompd/"
 
+import lit.llvm
+lit.llvm.initialize(lit_config, config)
+
 # Let the main config do the real work.
 lit_config.load_config(config, "@CMAKE_CURRENT_SOURCE_DIR@/lit.cfg")
diff --git a/openmp/runtime/test/lit.cfg b/openmp/runtime/test/lit.cfg
index cfbd2c5d418b5..ca32985fe6c48 100644
--- a/openmp/runtime/test/lit.cfg
+++ b/openmp/runtime/test/lit.cfg
@@ -70,7 +70,6 @@ if config.operating_system != 'Haiku':
         libs += " -latomic"
 
 # Allow REQUIRES / UNSUPPORTED / XFAIL to work
-config.target_triple = [ ]
 for feature in config.test_compiler_features:
     config.available_features.add(feature)
 
diff --git a/openmp/runtime/test/lit.site.cfg.in b/openmp/runtime/test/lit.site.cfg.in
index d0e54eac0dc3a..fc65289e4ce64 100644
--- a/openmp/runtime/test/lit.site.cfg.in
+++ b/openmp/runtime/test/lit.site.cfg.in
@@ -12,6 +12,7 @@ config.libomp_obj_root = "@CMAKE_CURRENT_BINARY_DIR@"
 config.library_dir = "@LIBOMP_LIBRARY_DIR@"
 config.omp_header_directory = "@LIBOMP_BINARY_DIR@/src"
 config.operating_system = "@CMAKE_SYSTEM_NAME@"
+config.target_triple = "@LLVM_TARGET_TRIPLE@"
 config.hwloc_library_dir = "@LIBOMP_HWLOC_LIBRARY_DIR@"
 config.using_hwloc = @LIBOMP_USE_HWLOC@
 config.has_ompt = @LIBOMP_OMPT_SUPPORT@ and @LIBOMP_OMPT_OPTIONAL@
@@ -24,5 +25,8 @@ config.target_arch = "@LIBOMP_ARCH@"
 config.compiler_frontend_variant = "@CMAKE_C_COMPILER_FRONTEND_VARIANT@"
 config.compiler_simulate_id = "@CMAKE_C_SIMULATE_ID@"
 
+import lit.llvm
+lit.llvm.initialize(lit_config, config)
+
 # Let the main config do the real work.
 lit_config.load_config(config, "@LIBOMP_BASE_DIR@/test/lit.cfg")
diff --git a/openmp/tools/archer/tests/lit.cfg b/openmp/tools/archer/tests/lit.cfg
index f8fbcad752a4c..c4bae61800ac4 100644
--- a/openmp/tools/archer/tests/lit.cfg
+++ b/openmp/tools/archer/tests/lit.cfg
@@ -58,7 +58,6 @@ if config.has_libatomic:
     libs += " -latomic"
 
 # Allow XFAIL to work
-config.target_triple = [ ]
 for feature in config.test_compiler_features:
     config.available_features.add(feature)
 
diff --git a/openmp/tools/archer/tests/lit.site.cfg.in b/openmp/tools/archer/tests/lit.site.cfg.in
index ddcb7b8bc3a56..dc7293972f083 100644
--- a/openmp/tools/archer/tests/lit.site.cfg.in
+++ b/openmp/tools/archer/tests/lit.site.cfg.in
@@ -11,11 +11,15 @@ config.libomp_obj_root = "@CMAKE_CURRENT_BINARY_DIR@"
 config.omp_library_dir = "@LIBOMP_LIBRARY_DIR@"
 config.omp_header_dir = "@LIBOMP_INCLUDE_DIR@"
 config.operating_system = "@CMAKE_SYSTEM_NAME@"
+config.target_triple = "@LLVM_TARGET_TRIPLE@"
 config.has_libatomic = @LIBARCHER_HAVE_LIBATOMIC@
 config.has_tsan = @OPENMP_TEST_ENABLE_TSAN@
 
 config.test_archer_flags = "@LIBARCHER_TEST_FLAGS@"
 config.libarcher_obj_root = "@CMAKE_CURRENT_BINARY_DIR@"
 
+import lit.llvm
+lit.llvm.initialize(lit_config, config)
+
 # Let the main config do the real work.
 lit_config.load_config(config, "@CMAKE_CURRENT_SOURCE_DIR@/lit.cfg")
diff --git a/openmp/tools/multiplex/tests/lit.cfg b/openmp/tools/multiplex/tests/lit.cfg
index 459250582ae95..24e7a5be4a8e0 100644
--- a/openmp/tools/multiplex/tests/lit.cfg
+++ b/openmp/tools/multiplex/tests/lit.cfg
@@ -63,7 +63,6 @@ if 'CHECK_OPENMP_ENV' in os.environ:
         config.environment[name] = value
 
 # Allow XFAIL to work
-config.target_triple = [ ]
 for feature in config.test_compiler_features:
     config.available_features.add(feature)
 
diff --git a/openmp/tools/multiplex/tests/lit.site.cfg.in b/openmp/tools/multiplex/tests/lit.site.cfg.in
index dbe7a33291eca..b6aff0de444e1 100644
--- a/openmp/tools/multiplex/tests/lit.site.cfg.in
+++ b/openmp/tools/multiplex/tests/lit.site.cfg.in
@@ -11,6 +11,10 @@ config.omp_library_dir = "@LIBOMP_LIBRARY_DIR@"
 config.omp_header_dir = "@LIBOMP_INCLUDE_DIR@"
 config.ompt_print_callback_dir = "@OMPT_PRINT_CALLBACKS_DIR@"
 config.operating_system = "@CMAKE_SYSTEM_NAME@"
+config.target_triple = "@LLVM_TARGET_TRIPLE@"
+
+import lit.llvm
+lit.llvm.initialize(lit_config, config)
 
 # Let the main config do the real work.
 lit_config.load_config(config, "@CMAKE_CURRENT_SOURCE_DIR@/lit.cfg")
