diff --git a/openmp/libompd/test/ompt_plugin.h b/openmp/libompd/test/ompt_plugin.h
index 4488e6884a06c..74cb625ecb375 100644
--- a/openmp/libompd/test/ompt_plugin.h
+++ b/openmp/libompd/test/ompt_plugin.h
@@ -127,8 +127,8 @@ static void ompt_finalize(ompt_data_t *tool_data) {}
 
 // "This func will be invoked by OpenMP implementation, refer spec: 4.2.1"
 // NOLINTNEXTLINE
-static ompt_start_tool_result_t *ompt_start_tool(unsigned int omp_version,
-                                                 const char *runtime_version) {
+ompt_start_tool_result_t *ompt_start_tool(unsigned int omp_version,
+                                          const char *runtime_version) {
   static ompt_start_tool_result_t ompt_start_tool_result = {
       &ompt_initialize, &ompt_finalize, {0}};
   return &ompt_start_tool_result;
